internal/tofu/transform_closeable_resource.go

```diff
@@ -36,67 +36,91 @@ func (t *CloseableResourceTransformer) Transform(_ context.Context, g *Graph) er
-			ncr = &nodeCloseableResource{
-				Addr: rn.ResourceAddr(),
-				cb:   rn.Close,
-			}
-			g.Add(ncr)
-			cpm[key] = ncr
+		_, exists := cpm[key]
+		if exists {
+			continue
 		}
 
-		// Close node depends on the resource itself
-		g.Connect(dag.BasicEdge(ncr, rn))
-
-		// We need to create a dependency between the closing node and the provider of the resource that we are closing.
-		// This is needed to ensure that the closing node will be added as a dependency later to the closing
-		// node of the provider.
-		for _, dn := range g.DownEdges(rn) {
-			switch dn.(type) {
-			case GraphNodeProvider:
-				g.Connect(dag.BasicEdge(ncr, dn))
+		// gather the references for the closing function from all existing instances of the resource
+		var callbacks []resourceCloser
+		var addr addrs.ConfigResource
+		var allDeps []dag.Vertex
+		for _, inst := range instances {
+			deps, err := t.collectInstanceDependencies(g, inst)
+			if err != nil {
+				return err
 			}
-
+			// collect all vertices inst is linked to
+			allDeps = append(allDeps, deps...)
+			// and collect inst as a dependency too, later it will be linked to the node responsible with closing it
+			allDeps = append(allDeps, inst)
+			callbacks = append(callbacks, inst.Close)
+			// Should be the same for all the instances since closeableResourcesVertexMap is generating the key
+			// based on each vertex addr. Therefore, no issue here on overwriting it on each iteration.
+			addr = inst.ResourceAddr()
 		}
-
-		// connect all the resource's dependencies to the close node to ensure that we are not executing the closing of the
-		// resource before having all the other references satisfied
-		desc, err := g.Descendents(rn)
-		if err != nil {
-			return err
+		// we postponed creation of the node and its addition to the graph, just to ensure that we are having all the
+		// required information prepared without errors before adding this node into the graph.
+		ncr := &nodeCloseableResource{
+			Addr: addr,
+			cbs:  callbacks,
 		}
-		for _, s := range desc {
-			switch s.(type) {
-			case GraphNodeReferencer:
-				g.Connect(dag.BasicEdge(ncr, s))
-			}
+		g.Add(ncr)
+		for _, dep := range allDeps {
+			g.Connect(dag.BasicEdge(ncr, dep))
 		}
+		cpm[key] = struct{}{}
 	}
 	return nil
 }
 
+func (t *CloseableResourceTransformer) collectInstanceDependencies(g *Graph, inst closableResource) ([]dag.Vertex, error) {
+	var deps []dag.Vertex
+	// We need to create a dependency between the closing node and the provider of the resource that we are closing.
+	// This is needed to ensure that the closing node will be added as a dependency later to the closing
+	// node of the provider.
+	for _, dn := range g.DownEdges(inst) {
+		switch dn.(type) {
+		case GraphNodeProvider:
+			deps = append(deps, dn)
+			//g.Connect(dag.BasicEdge(ncr, dn))
+		}
+
+	}
+
+	// connect all the resource's dependencies to the close node to ensure that we are not executing the closing of the
```

Comment:
Old comment paired with the commented out connect below?

Reply:
Thanks a lot. Ugly slip up on my side. Fixed in a096347897b66a03f32c11392caf8ce8b7a18ba2